### ISSUE TITLE
Demangling of Microsoft-specific symbols doesn't work

### ISSUE THREAD

**Issue by dennisbabkin** [NationalSecurityAgency/ghidra#438]
Unless I'm missing some setting, the demangling (un-decoration) of the latest Microsoft-specific symbols doesn't seem to work.

I'm testing it with Ghidra v.9.0.2

Example: While de-compiling a function from cmctrl32.dll (x86, v.6.10.17134.286 on my Windows 10) almost all Microsoft symbols are not demangled:

![2_2](https://user-images.githubusercontent.com/25473659/56195993-db6f4800-5fea-11e9-9393-a856cece8454.PNG)

I'm not sure if the current implementation uses [UnDecorateSymbolName](https://docs.microsoft.com/en-us/windows/desktop/api/dbghelp/nf-dbghelp-undecoratesymbolname) function or not. If so, it is **old** and **doesn't work well** with most modern mangling schemes that Microsoft employs in Visual Studio compiler.

I did this research for my own tool that seems to demangle such symbols without a problem. So let me share it. Here's a screenshot of the same symbol that Ghidra failed to demangle:

![2_1](https://user-images.githubusercontent.com/25473659/56196321-76682200-5feb-11e9-907f-457315a4e1d2.PNG)

So it would be nice if that is corrected as copying and pasting those names back and forth gets quite tedious.

----------------------------------------

## My solution

Use undocumented but exportable function from `vcruntime140.dll` called `__unDNameEx`. (It beats me why Microsoft doesn't want to document it.)

I am not good at Java, so instead let me show the C++ implementation (so hopefully someone can translate it into Java).

To make it safer (while dealing with "undocumented" function) I would statically link the module that uses it to VC RT v.14. After that you can declare that function as a static import as such:

	//Import this undocumented function from the vcruntime140.dll (if we're statically linking to it)
	extern "C" PSTR __cdecl __unDNameEx(
		PSTR buffer, 
		PCSTR mangled, 
		DWORD cb,
		void* (__cdecl* memget)(DWORD),
		void (__cdecl* memfree)(void*),
		PSTR (__cdecl* GetParameter)(long i),
		DWORD flags
	);

Otherwise, if static linking is not possible, you will have to load it at runtime:

	typedef PSTR __cdecl UNDEC_NAME_EX(
		PSTR buffer, 
		PCSTR mangled, 
		DWORD cb,
		void* (__cdecl* memget)(DWORD),
		void (__cdecl* memfree)(void*),
		PSTR (__cdecl* GetParameter)(long i),
		DWORD flags
	);

	HMODULE hModVcRT140 = ::LoadLibrary(L"vcruntime140.dll");
	UNDEC_NAME_EX *__unDNameEx = (UNDEC_NAME_EX*)::GetProcAddress(hModVcRT140, "__unDNameEx");

	if(__unDNameEx)
	{
		//Do work here ....

	}

Obviously remember to free the module when done:

	::FreeLibrary(hModVcRT140);

Lastly use that function to demangle Microsoft-specific symbols, as such:

	//IMPORTANT: This function is not multi-threading safe. If calling it from
	//           multiple threads, do so from under a critical section.

	const static char p_strNameToUndecorate[] = "?_SBGetParms@@YGHPAUHWND__@@HPBUtagSBDATA@@PAUtagSCROLLINFO@@@Z";

	//Flags for conversion (see documented UnDecorateSymbolName() function for details)
	//eg: UNDNAME_COMPLETE or UNDNAME_NO_MS_KEYWORDS, etc.
	DWORD dwFlags = UNDNAME_COMPLETE;

	//Undecorate now (let it reserve its own memory by passing NULL as a first parameter)
	char* pUndecorated = __unDNameEx(NULL, p_strNameToUndecorate, 0, _funcAlloc, _funcFree, _funcGetParameter, dwFlags);
	if(pUndecorated)
	{
		//All good
		printf("Result: %s\n", pUndecorated);

		//Free mem
		delete[] pUndecorated;
		pUndecorated = NULL;
	}
	else
	{
		//Failed
		printf("FAILED\n");
	}

This function doesn't come in the wide-string (or Unicode-16) flavor so you'll have to convert such strings to ANSI first.

Lastly, use these callbacks for memory management inside that function:

	static void* __cdecl _funcAlloc(ULONG cb)
	{
		//Reserve mem for 'cb' chars
		return new (std::nothrow) char[cb];
	}

	static void __cdecl _funcFree(void* p)
	{
		//Free memory in 'p'
		if(p)
			delete[] p;
	}

	static PSTR __cdecl _funcGetParameter(long i)
	{
		static char dummy[] = "";
		return dummy;
	}

**Comment by emteere** [comment]
If you delete the function, then place your cursor on the mangled label, and run the script DemangleSymbolScript, what happens?

**Comment by dennisbabkin** [comment]
@emteere in that case it seems to demangle it correctly for one function I tried it with. Then the issue is how do you incorporate it into the static analysis to do it automatically on all symbols? (Again, am I missing the setting?)

**Comment by emteere** [comment]
You can use the DemangleAllScript, or you can re-analyze the program.
There is already an automated Demangler analyzer, however I suspect your symbols were loaded from a .PDB file after the fact.  If the symbols were applied by loadding the .PDB, the demangler may not have been kicked off again, although I think it should have been kicked off.  I'd need to know more about exactly how you loaded the binary and when you applied the .PDB symbols.
If you were to re-import the binary with the .PDB already downloaded and ready to apply, then I believe it would have demangled automatically.

**Comment by emteere** [comment]
The msdia.dll is used to load the PDB.  Ghidra has it's own internal demangler.  We try not to rely on software that is only available on one platform.

**Comment by dennisbabkin** [comment]
@emteere yeah, I admit I wasn't sure how to load the symbols. It's somewhat counter-intuitive the way it does it. I think I went to File -> Download PDB file and downloaded it from the Microsoft symbol server. After that it offered to apply it to my project, which I did.

So instead of me guessing, what is the correct sequence of applying PDB symbols from an online symbol server?

And also how can I apply them to a file I've been already working on? I'd hate to lose several days of RE work on it.

Thanks.

PS. As for having its own internal demangler, it may be both good and bad. As you probably know Microsoft doesn't really stick to standards. In my own project that I listed above I had to literally RE how they do it in their debuggers because their documented demangling API wasn't doing a good job. So after I find a way to load all the symbols automatically in Ghidra I can report back to you about the results. So maybe your own internal demangler does a good job. At this point I don't know.

**Comment by Zhentar** [comment]
http://msdl.microsoft.com/download/symbols/ucrtbased.dll/C09C8E211C2000/ucrtbased.dll contains quite a few symbols that don't demangle correctly. Conveniently, it also contains the correct logic for demangling them.

**Comment by dennisbabkin** [comment]
@Zhentar possibly. What I provided in my original post is the source code to do it in general. (Obviously for a Windows platform.) If you want to test it in action here's the [link](https://dennisbabkin.com/utilities/#WinApiSearch) to the tool in my original screenshot that implements that `__unDNameEx` call from the VC run time library. Go to View -> Undecorate C++ Symbols. That API is totally undocumented, but it seems to demangle most of the latest Microsoft symbols.
